// render-jsx/dist/es5/component/plugins/live-component.processor.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.LiveComponentProcessor = void 0;
var processor_1 = require("../processor");
var LiveComponentProcessor = /** @class */ (function (_super) {
    __extends(LiveComponentProcessor, _super);
    function LiveComponentProcessor() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    LiveComponentProcessor.prototype.process = function (provide, post) {
        var renderer = this.renderer();
        var hooks = [];
        provide({
            onBind: function (fn) { return hooks.push({ bind: fn }); },
            onClear: function (fn) { return hooks.push({ clear: fn }); },
        });
        post(function (node) {
            hooks.forEach(function (hook) { return renderer.hook(node, hook); });
        });
    };
    LiveComponentProcessor.prototype.priority = function () {
        return processor_1.ComponentProcessor.PriorityFallback;
    };
    return LiveComponentProcessor;
}(processor_1.ComponentProcessor));
exports.LiveComponentProcessor = LiveComponentProcessor;
