// render-jsx/dist/es5/dom/util/life-cycle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.attachLifeCycleHook = exports.lifeCycleBind = exports.lifeCycleClear = exports.getLifeCycleMarker = exports.setLifeCycleMarker = exports.fragmentLifeCycleInfo = exports.lifeCycleInfo = void 0;
function lifeCycleInfo(node, createIfNonExistent) {
    if (createIfNonExistent === void 0) { createIfNonExistent = false; }
    var _node = node;
    if (node.nodeType === node.DOCUMENT_FRAGMENT_NODE) {
        return fragmentLifeCycleInfo(node, createIfNonExistent);
    }
    else {
        if (_node.lifecycle) {
            return _node.lifecycle;
        }
        else if (createIfNonExistent) {
            _node.lifecycle = { bound: false };
            return _node.lifecycle;
        }
    }
}
exports.lifeCycleInfo = lifeCycleInfo;
function fragmentLifeCycleInfo(fragment, createIfNonExistent) {
    var _a;
    var marker = getLifeCycleMarker(fragment);
    /*istanbul ignore if*/
    if (marker) {
        return lifeCycleInfo(marker, createIfNonExistent);
    }
    else if (createIfNonExistent) {
        marker = (_a = fragment.ownerDocument) === null || _a === void 0 ? void 0 : _a.createTextNode('');
        setLifeCycleMarker(fragment, marker);
        return lifeCycleInfo(marker, true);
    }
}
exports.fragmentLifeCycleInfo = fragmentLifeCycleInfo;
function setLifeCycleMarker(fragment, marker) {
    fragment.lifecycleMarker = marker;
    if (!fragment.contains(marker)) {
        fragment.appendChild(marker);
    }
}
exports.setLifeCycleMarker = setLifeCycleMarker;
function getLifeCycleMarker(fragment) {
    return fragment.lifecycleMarker;
}
exports.getLifeCycleMarker = getLifeCycleMarker;
function lifeCycleClear(node) {
    var lifecycle = lifeCycleInfo(node);
    if (lifecycle) {
        /*istanbul ignore next*/
        if (lifecycle.cleared) {
            return;
        }
        lifecycle.cleared = true;
        if (lifecycle.clears) {
            for (var i = 0, clear = lifecycle.clears[i]; i < lifecycle.clears.length; clear = lifecycle.clears[++i]) {
                clear();
            }
            lifecycle.clears = undefined;
        }
    }
    var children = node.childNodes;
    for (var i = 0, child = children.item(i); i < children.length; child = children.item(++i)) {
        lifeCycleClear(child);
    }
}
exports.lifeCycleClear = lifeCycleClear;
function lifeCycleBind(node) {
    var lifecycle = lifeCycleInfo(node);
    if (lifecycle) {
        /*istanbul ignore next*/
        if (lifecycle.bound) {
            return;
        }
        lifecycle.bound = true;
        if (lifecycle.binds) {
            for (var i = 0, bind = lifecycle.binds[i]; i < lifecycle.binds.length; bind = lifecycle.binds[++i]) {
                bind();
            }
        }
        lifecycle.binds = undefined;
    }
    var children = node.childNodes;
    for (var i = 0, child = children.item(i); i < children.length; child = children.item(++i)) {
        lifeCycleBind(child);
    }
}
exports.lifeCycleBind = lifeCycleBind;
function attachLifeCycleHook(hook, node) {
    var lifecycle = lifeCycleInfo(node, true);
    if (hook.bind) {
        (lifecycle.binds || (lifecycle.binds = [])).push(hook.bind);
    }
    if (hook.clear) {
        (lifecycle.clears || (lifecycle.clears = [])).push(hook.clear);
    }
}
exports.attachLifeCycleHook = attachLifeCycleHook;
